n8n-io/n8n
# fix(core): Fix removal of triggers and pollers from memory on deactivation in multi-main setup

On leadership transition, we call the multi-main-specific method `ActiveWorkflows.removeAllTriggerAndPollerBasedWorkflows`, which stops all triggers and pollers, but this method does _not_ deregister the workflow from memory, as done in `ActiveWorkflows.close`. To fix this, correct this oversight and also [avoid using `async` with `forEach`](https://gist.github.com/joeytwiddle/37d2085425c049629b80956d3c618971).

To test: 
- Launch a worker and two mains and call `http://localhost:{port}/rest/debug/multi-main-setup` to see the state on both pre-transition. Leader should have triggers and pollers; follower should not.
- Delete the leader key with `docker exec n8n-redis redis-cli DEL "n8n:main_instance_leader"` and see which instance becomes leader again. Whichever gained leadership, the debug endpoint should report that only that new leader has triggers and pollers. Make sure to test both scenarios.

This test fails on `master` - after leadership transition, the debug endpoint reports both new and former leader with active triggers. We were not seeing this because we were trusting the logs.
